proposals/route-blinding.md

```diff
@@ -0,0 +1,308 @@
+    B(i) = HMAC256("blinded_node_id", ss(i)) * P(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+
+Blinded route:
+  
+  (P(0),fees(0),cltv(0),encrypted_data(0))
+  (B(1),fees(1),cltv(1),encrypted_data(1))
+  ...
+  (B(r-1),fees(r-1),cltv(r-1),encrypted_data(r-1))
+```
+
+Note that this is exactly the same construction as Sphinx, but at each hop we use the shared secret
+to derive a blinded `node_id` for `N(i)` for which the private key will only be known by `N(i)`.
+
+The recipient needs to provide `E(0)` and the blinded route to potential senders.
+
+The `encrypted_data(i)` is encrypted with ChaCha20-Poly1305 using the `rho(i)` key, and
+contains the real `short_channel_id` to forward to (and potentially other fields).
+
+Note that the introduction point uses the real `node_id`, not the blinded one, because the sender
+needs to be able to locate this introduction point and find a route to it. But the sender will send
+`E(0)` in the onion `hop_payload` for `N(0)`, which will allow the introduction point to compute
+the shared secret and correctly forward.
+
+Note that in the specific case of payments, the recipient can sign the invoice with `e(0)`.
+The sender will recover `E(0)` from the signature so no extra field needs to be added to Bolt 11.
+And this ensures the recipient doesn't reveal his real `node_id` through the invoice signature.
+
+However, if the recipient wants to be able to prove invoice ownership in the future, she should
+sign the invoice with a different key and provide `E(0)` via a dedicated Bolt 11 field.
+
+### Sending to a blinded route
+
+The sender finds a route to the introduction point `N(0)`, and extends it with the blinded route.
+It then creates an onion for that route, and includes `E(0)` and `encrypted_data(0)` in
+the onion payload for `N(0)`.
+
+When `N(0)` receives the onion and decrypts it, it finds `E(0)` in the payload and is able to
+compute the following:
+
+```text
+  ss(0) = H(k(0) * E(0))
+  rho(0) = HMAC256("rho", ss(0))
+  E(1) = H(E(0) || ss(0)) * E(0)
+```
+
+It uses `rho(0)` to decrypt the `encrypted_data(0)` and discover the `scid` to forward to.
+It forwards the onion to the next node and includes `E(1)` in a TLV field in the message extension
+(at the end of the `update_add_htlc` message).
+
+All the following intermediate nodes `N(i)` do the following steps:
+
+```text
+  E(i) <- extracted from update_add_htlc's TLV extension
+  ss(i) = H(k(i) * E(i))
+  b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)
+  Use b(i) to decrypt the incoming onion
+  rho(i) = HMAC256("rho", ss(i))
+  Use rho(i) to decrypt the `encrypted_data` inside the onion and discover the next node
```

Comment:
I will leave the above comments but I think I finally realized why I was confused. you don't use the blinded `b(i)` to decrypt the `encrypted_data`. I think I was confused because `rho` is also used in the regular onion scheme 

Just to be clear `B(i)` for `i >0` and `P(0)` are being used to encrypt the onions where as the `rho(i)` are being used to encrypt the `encrypted_data`.  Thus you need to communicate the blinded `node_id`s because `N(i)`s  could not decrypt the onions with `k(i)`.

Reply:
Yes, that's exactly it. I felt it made sense to separate these concerns, as they happen at different layers.